docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
+The main difference with the current approach are:
 
+* the image spin is selected depending on the OS version
+* all language dependencies are installed at build time
+* languages not offereing binaries are pre-compiled by Read the Docs and stored in the cache
+* miniconda/mambaforge are now managed with the same management tool
 
-Specifying extra user's dependencies
-------------------------------------
 
-Different users may have different requirements. We were already requested to install
-``swig``, ``imagemagick``, ``libmysqlclient-dev``, ``lmod``, ``rust``, ``poppler-utils``, etc.
+Specifying extra languages requirements
+---------------------------------------
 
-People with specific dependencies will be able to install them as APT packages or as extras
-using ``.readthedocs.yaml`` config file. Example:
+Different users may have different requirements.
+People with specific language dependencies will be able to install them by using ``.readthedocs.yaml`` config file.
+Example:
 
 .. code:: yaml
 
+   version: 3
    build:
-     image: ubuntu20
-     python: 3.9
-     system_packages:
-       - swig
-       - imagemagick
-     extras:
-       - node==14
-       - rust==1.46
+     os: ubuntu20
+     languages:
+       python: "3.9"  # supports "pypy3", "miniconda3" and "mambaforge"
+       nodejs: "14"
+       rust: "1.54.0"
+       golang: "1.17"
 
 Important highlights:
 
-* users won't be able to use custom Ubuntu PPAs to install packages
-* all APT packages installed will be from official Ubuntu repositories
-* not specifying ``build.image`` will pick the latest OS image available
-* not specifying ``build.python`` will pick the latest Python version available
-* Ubuntu 18 will still be available via ``stable`` and ``latest`` images
-* all ``node`` (major) pre-compiled versions on ``nodenv`` are available to select
-* all ``rust`` (minor) pre-compiled versions on ``rustup`` are available to select
-* knowing exactly what packages users are installing,
-  could allow us to prebuild extra images: ``ubuntu20-py37+node14``
+* do not treat Python language different from the others (will help us to support other non-Python doctools in the future)
+* specifying ``build.languages.python: "3"`` will use Python version ``3.x.y``, and may differ between builds
+* specifying ``build.languages.python: "3.9"`` will use Python version ``3.9.y``, and may differ between builds
+* specifying ``build.languages.nodejs: "14"`` will use nodejs version ``14.x.y``, and may differ between builds
+* if no full version is declared, the latest available on ``asdf`` will be installed
+* specifying minor language versios is not allowed (e.g. ``3.7.11``)
+* not specifying ``build.os`` will make the config file parser to fail
+* not specifying ``build.languages`` will make the config file parsing to fail (at least one is required)
+* specifying only ``build.languages.nodejs`` and using Sphinx to build the docs, will make the build to fail
```

Comment:
The build will fail, but the config will parse?

Reply:
The config will parse correctly, yes. We can't be 100% sure what's the doctool the user is using to build the docs. We may be to infer it from other configs, but I'd avoid doing this guessing for now.

I don't want to _require_ `python` as a language here, because in the future we may support other doctools that don't depend on Python at all.